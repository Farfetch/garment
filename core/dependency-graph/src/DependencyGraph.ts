function createGraphWalker<T>(
  edges: Map<T, Set<T>>,
  options: { includeInitialNode?: boolean } = {}
) {
  const { includeInitialNode } = options;

  const result = new Set<T>();
  const currentPath = new Set<T>();
  const initialNodes = new Set();

  function graphWalker(node: T, depth = 0) {
    currentPath.add(node);

    if (depth === 0) {
      initialNodes.add(node);
    }
    if (result.has(node)) {
      result.delete(node);
    }
    if (!initialNodes.has(node) || includeInitialNode) {
      result.add(node);
    }

    const nodeEdges = edges.get(node);
    if (nodeEdges) {
      for (const nodeEdge of nodeEdges.values()) {
        if (!currentPath.has(nodeEdge)) {
          graphWalker(nodeEdge, depth + 1);
        } else {
          throw new Error(
            'Graph contains circular dependency: ' +
              [...currentPath, nodeEdge].join(' -> ')
          );
        }
      }
      currentPath.delete(node);
    }
  }

  return (...nodes: T[]) => {
    for (const node of nodes) {
      graphWalker(node);
    }
    return result;
  };
}

function arrayfy<T>(arr: T | T[]) {
  return Array.isArray(arr) ? arr : [arr];
}

export class Graph<T = string> {
  private nodes = new Set<T>();
  private outEdges = new Map<T, Set<T>>();
  private inEdges = new Map<T, Set<T>>();

  [Symbol.iterator]() {
    return this.nodes[Symbol.iterator]();
  }

  get size() {
    return this.nodes.size;
  }

  addNode(...nodes: T[]) {
    for (const node of nodes) {
      if (!this.nodes.has(node)) {
        this.nodes.add(node);
        this.outEdges.set(node, new Set());
        this.inEdges.set(node, new Set());
      }
    }
  }

  removeNode(...nodes: T[]) {
    for (const node of nodes) {
      if (this.nodes.has(node)) {
        this.nodes.delete(node);
        this.inEdges.delete(node);
        this.outEdges.delete(node);
        [...this.inEdges.values(), ...this.outEdges.values()].forEach(edges => {
          if (edges.has(node)) {
            edges.delete(node);
          }
        });
      }
    }
  }

  hasNode(node: T) {
    return this.nodes.has(node);
  }

  addDependency(from: T, to: T) {
    if (from === to) {
      throw new Error(`Node can't be dependant on itself: ${from}`);
    }
    if (!this.hasNode(from)) {
      throw new Error('Node does not exist: ' + from);
    }
    if (!this.hasNode(to)) {
      throw new Error('Node does not exist: ' + to);
    }
    const outEdges = this.outEdges.get(from);
    const inEdges = this.inEdges.get(to);
    if (!outEdges || !inEdges) {
      return;
    }

    if (!outEdges.has(to)) {
      outEdges.add(to);
    }
    if (!inEdges.has(from)) {
      inEdges.add(from);
    }
  }

  removeDependency(from: T, to: T) {
    if (!this.hasNode(from)) {
      throw new Error('Node does not exist: ' + from);
    }
    if (!this.hasNode(to)) {
      throw new Error('Node does not exist: ' + to);
    }

    const outEdges = this.outEdges.get(from);
    const inEdges = this.inEdges.get(to);
    if (!outEdges || !inEdges) {
      return;
    }

    if (outEdges.has(to)) {
      outEdges.delete(to);
    }
    if (inEdges.has(from)) {
      inEdges.delete(from);
    }
  }

  getDependenciesOf(...nodes: T[]) {
    const walk = createGraphWalker(this.outEdges);
    return Array.from(walk(...nodes));
  }

  getDirectDependenciesOf(node: T) {
    return Array.from(this.outEdges.get(node) || []);
  }

  getDependantsOf(...nodes: T[]) {
    const walk = createGraphWalker(this.inEdges);
    return Array.from(walk(...nodes));
  }

  getDirectDependantsOf(node: T) {
    return Array.from(this.inEdges.get(node) || []);
  }

  getNodesWithoutDependants() {
    const nodes = new Set<T>();
    for (const [node, egdes] of this.inEdges) {
      if (egdes.size === 0) {
        nodes.add(node);
      }
    }
    return Array.from(nodes);
  }

  getLeafNodes() {
    const nodes = new Set<T>();
    for (const [node, egdes] of this.outEdges) {
      if (egdes.size === 0) {
        nodes.add(node);
      }
    }
    return Array.from(nodes);
  }

  getOverallOrder() {
    return this.sort([...this.nodes]);
  }

  sort(nodes: T[]) {
    const walk = createGraphWalker(this.outEdges, { includeInitialNode: true });
    return Array.from(walk(...nodes))
      .filter(node => nodes.includes(node))
      .reverse();
  }

  map<O>(fn: (node: T) => O | O[]): Graph<O> {
    const graph = new Graph<O>();

    const nodeMapping = new Map<T, O[]>();
    const mapNode = (node: T) => nodeMapping.get(node)!;
    const setify = <P>(arr: P[][]) =>
      arr.reduce(
        (result, item) => (item.forEach(i => result.add(i)), result),
        new Set<P>()
      );

    for (const node of this.nodes) {
      const mapped = arrayfy(fn(node));
      nodeMapping.set(node, mapped);
      mapped.forEach(node => graph.nodes.add(node));
    }
    for (const [key, nodeSet] of this.inEdges) {
      mapNode(key).forEach(keyNode => {
        graph.inEdges.set(keyNode, setify([...nodeSet].map(mapNode)));
      });
    }
    for (const [key, nodeSet] of this.outEdges) {
      mapNode(key).forEach(keyNode => {
        graph.outEdges.set(keyNode, setify([...nodeSet].map(mapNode)));
      });
    }

    return graph;
  }

  filter(fn: (node: T) => boolean): Graph<T> {
    const graph = new Graph<T>();

    const shouldInclude = new Set<T>();

    for (const node of this) {
      if (fn(node)) {
        shouldInclude.add(node);
        graph.nodes.add(node);
      }
    }

    for (const [key, inNodes] of this.inEdges) {
      if (shouldInclude.has(key)) {
        graph.inEdges.set(
          key,
          new Set(
            Array.from(inNodes).filter(inNode => shouldInclude.has(inNode))
          )
        );
      }
    }

    for (const [key, outNodes] of this.outEdges) {
      if (shouldInclude.has(key)) {
        graph.outEdges.set(
          key,
          new Set(
            Array.from(outNodes).filter(outNode => shouldInclude.has(outNode))
          )
        );
      }
    }

    return graph;
  }

  clone() {
    const clone = new Graph<T>();
    for (const node of this.nodes) {
      clone.nodes.add(node);
    }
    for (const [key, value] of this.inEdges) {
      clone.inEdges.set(key, new Set([...value]));
    }
    for (const [key, value] of this.outEdges) {
      clone.outEdges.set(key, new Set([...value]));
    }
    return clone;
  }

  assign(...graphs: Graph<T>[]) {
    for (const graph of graphs) {
      for (const node of graph) {
        this.addNode(node);
      }
      for (const [key, nodes] of graph.inEdges) {
        this.inEdges.set(key, new Set(nodes));
      }
      for (const [key, nodes] of graph.outEdges) {
        this.outEdges.set(key, new Set(nodes));
      }
    }
  }

  getIndependentSubgraphs(): Graph<T>[] {
    const sets: Set<T>[] = [];

    const topNodes = this.getNodesWithoutDependants();
    for (const topNode of topNodes) {
      const deps = [topNode, ...this.getDependenciesOf(topNode)];
      if (sets.length) {
        let addedToExisting = false;
        for (const nodeSet of sets) {
          if (deps.some(node => nodeSet.has(node))) {
            deps.forEach(node => nodeSet.add(node));
            addedToExisting = true;
            break;
          }
        }
        if (!addedToExisting) {
          sets.push(new Set(deps));
        }
      } else {
        sets.push(new Set(deps));
      }
    }
    return sets.map(nodeSet => this.filter(node => nodeSet.has(node)));
  }

  async traverseParallel(
    cb: (node: T) => Promise<void> | void,
    limit = Infinity
  ) {
    const cloned = this.clone();

    await new Promise((resolve, reject) => {
      const inProgress = new Set<T>();
      let shouldStop = false;
      const iterate = () => {
        const leafNodes = shouldStop ? [] : cloned.getLeafNodes();
        if (!leafNodes.length && !inProgress.size) {
          return shouldStop ? reject() : resolve();
        }
        const nodesToExecute = [...leafNodes]
          .filter(node => !inProgress.has(node))
          .slice(0, limit - inProgress.size);
        if (nodesToExecute.length) {
          nodesToExecute.forEach(async node => {
            inProgress.add(node);

            try {
              await cb(node);

              cloned.removeNode(node);
            } catch (error) {
              shouldStop = true;
            }

            inProgress.delete(node);

            iterate();
          });
        }
      };
      iterate();
    });
  }

  cleanUnnecessaryDependencies() {
    for (const node of this) {
      const directDependencies = new Set(this.getDirectDependenciesOf(node));
      if (directDependencies.size) {
        directDependencies.forEach(directDependency => {
          const indirectDependencies = this.getDependenciesOf(directDependency);
          indirectDependencies.forEach(indirectDepNode => {
            if (directDependencies.has(indirectDepNode)) {
              this.removeDependency(node, indirectDepNode);
              directDependencies.delete(indirectDepNode);
            }
          });
        });
      }
    }
  }

  toString() {
    return this.getOverallOrder().join(', ');
  }
}
