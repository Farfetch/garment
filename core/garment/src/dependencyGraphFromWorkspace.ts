import { Project, Workspace } from '@garment/workspace';
import { Graph } from '@garment/dependency-graph';

export function dependencyGraphFromWorkspace(workspace: Workspace) {
  const dependencyGraph = new Graph<Project>();

  workspace.projects
    .list()
    .forEach(project => dependencyGraph.addNode(project));

  // From garment.json
  workspace.projects.list().forEach(project => {
    project.dependecies.forEach(depName => {
      const depProject = workspace.projects.get(depName);
      if (depProject) {
        dependencyGraph.addDependency(project, depProject);
      } else {
        throw new Error(
          `Can't find dependency of the project "${project.name}" called "${depName}"`
        );
      }
    });
  });

  // From package.json
  const packageToProjectMap = new Map<string, Project>();
  const nodeProjects = workspace.projects
    .list()
    .filter(project => project.nodePackage)
    .map(project => {
      return {
        name: project.name,
        project,
        nodePackage: project.nodePackage!
      };
    });

  for (const { project, nodePackage } of nodeProjects) {
    packageToProjectMap.set(nodePackage.name, project);
  }

  for (const { project, nodePackage } of nodeProjects) {
    for (const dep of nodePackage.allDependencies.keys()) {
      const depProjectName = packageToProjectMap.get(dep);
      if (depProjectName) {
        dependencyGraph.addDependency(project, depProjectName);
      }
    }
  }

  return dependencyGraph;
}
