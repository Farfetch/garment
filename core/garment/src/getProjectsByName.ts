import { Project, Workspace } from '@garment/workspace';
import * as matcher from 'matcher';

import isValidPath = require('is-valid-path');

export function getProjectsByName(
  workspace: Workspace,
  projectNames: string[],
  files?: boolean
) {
  const { projects } = workspace;
  let result: Project[] = [];
  const filesByProject = new Map<Project, string[]>();
  if (files) {
    projectNames.forEach(path => {
      const project = projects.getByPath(path);
      if (!project) {
        throw new Error(`Project containing file "${path}" was not found`);
      }
      if (!result.includes(project)) {
        result.push(project);
        if (!filesByProject.has(project)) {
          filesByProject.set(project, []);
        }
        filesByProject.get(project)!.push(path);
      }
    });
  } else {
    const workspaceProjectNames = projects.projectNames;
    projectNames.forEach(projectName => {
      if (projectName.search(/[*!]/) !== -1) {
        const matchedProjects = matcher(workspaceProjectNames, [projectName]);
        if (matchedProjects.length) {
          result.push(...matchedProjects.map(_ => projects.get(_)!));
        } else {
          throw new Error(`Projects matching "${projectName}" were not found`);
        }
      } else {
        let project = projects.get(projectName);
        if (!project && isValidPath(projectName)) {
          project = projects.getByPathExact(workspace.resolvePath(projectName));
        }
        if (project) {
          result.push(project);
        } else {
          throw new Error(`Project with path "${projectName}" was not found`);
        }
      }
    });
  }
  return result.map(project =>
    filesByProject.has(project)
      ? { project, files: filesByProject.get(project)! }
      : project
  );
}
