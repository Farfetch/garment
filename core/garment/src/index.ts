export * from './garment';
export * from './dependencyGraphFromWorkspace';
export * from './getProjectsByName';
