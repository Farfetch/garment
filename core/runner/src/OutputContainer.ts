import { Level, Logger } from '@garment/logger';
import { File } from './types';
import { createHash } from 'crypto';
import * as Path from 'path';

export interface CacheProvider<T = any> {
  has(id: string): Promise<boolean> | boolean;
  get(id: string): Promise<T> | T;
  set(id: string, content: T): Promise<void> | void;
}
export interface LogEntry {
  level: Level;
  content: any[];
  scope: string;
}

export class OutputContainer {
  static isOutputContainer(obj: any): obj is OutputContainer {
    return (<OutputContainer>obj)?.isOutputContainer ?? false;
  }

  private readonly isOutputContainer = true;

  readonly logger: Logger;
  readonly logs: LogEntry[] = [];
  readonly dependencies: string[];
  readonly files: File[] = [];
  readonly hash: string;
  readonly target: string;

  targetBaseDir: string;

  constructor(
    public readonly cacheProvider: CacheProvider,
    target: string | File,
    public readonly cacheKeys: string[],
    dependencies: (string | File)[] = []
  ) {
    const hash = createHash('md5');
    for (const cacheKey of this.cacheKeys) {
      hash.update(JSON.stringify(cacheKey));
    }
    this.hash = hash.digest('hex');

    if (typeof target === 'string') {
      this.target = target;
      this.targetBaseDir = Path.dirname(target);
    } else {
      this.target = target.absolutePath ?? target.path;
      this.targetBaseDir = target.baseDir ?? Path.dirname(this.target);
    }

    this.dependencies = dependencies.map(dep =>
      typeof dep === 'string' ? dep : dep.absolutePath ?? dep.path
    );

    this.logger = new Logger('', '', 'silly');
    this.logger.interceptors.push(({ level, args }) => {
      this.logs.push({
        level,
        content: args,
        scope: '    '
      });
      return false; // creating logger instance which doesn't output to the console
    });
  }

  add(file: File) {
    this.files.push(file);
  }

  get isNotCached() {
    return !this.cacheProvider.has(this.hash);
  }

  toJSON() {
    const { cacheKeys, dependencies, logs, files } = this;
    return {
      cacheKeys,
      dependencies,
      logs,
      files
    };
  }
}
