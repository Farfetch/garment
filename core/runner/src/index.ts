export * from './types';
export * from './getRunnerContext';
export * from './defineRunner';
export * from './getRunnerMeta';
export * from './OutputContainer';
