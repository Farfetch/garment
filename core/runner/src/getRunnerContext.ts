import { Logger } from '@garment/logger';
import { Project, Workspace } from '@garment/workspace';
import * as originalFs from 'fs';
import * as Mustache from 'mustache';
import { CacheProvider, OutputContainer } from './OutputContainer';
import { File, JsonFile, TextFile } from './types';
import * as Path from 'path';

(Mustache as any).escape = (text: string) => text;

export type GenericContext = ReturnType<typeof getRunnerContext>;
export type ContextOptions<O> = O & { outputDir?: string };

type GeneratorNextType<T> = T extends () => Generator<infer R, any, any>
  ? R
  : never;

export type Batch<O> = GeneratorNextType<GenericContext['batch']> & {
  options: ContextOptions<O>;
};

export type BatchGenerator<O> = () => Generator<Batch<O>>;

export interface Context<O> extends GenericContext {
  options: ContextOptions<O>;
  batch: BatchGenerator<O>;
}

export type InputFnCallbackReturn =
  | File
  | OutputContainer
  | File[]
  | OutputContainer[]
  | void;

export type InputFnCallBack<T = File> = (
  arg: T
) => InputFnCallbackReturn | Promise<InputFnCallbackReturn>;

export interface InputFn {
  (fn: InputFnCallBack<File[]>): void;
  forEach(fn: InputFnCallBack): void;
}

export interface GetRunnerContextOptions<O> {
  workspace: Workspace;
  project: Project;
  options: O;
  input: InputFn;
  fs?: typeof originalFs;
  dependsOnFile?(...paths: string[]): void;
  watchDependencies?(): void;
  longRunning?(onDestroy?: () => void): void;
  commitFilesToFS?(): Promise<void>;
  cacheProvider: CacheProvider;
  logger: Logger;
  defaultCacheKeys?: string[];
  batch?: { project: Project; options: O }[];
  outputDir?: string;
}

export function file(file: File) {
  return file;
}
file.text = (path: string, data: string, additional?: Partial<TextFile>) =>
  file({ type: 'text', path, data, ...additional }) as TextFile;
file.json = (path: string, data: object, additional?: Partial<JsonFile>) =>
  file({ type: 'json', path, data, ...additional }) as JsonFile;

export function renderTemplate(template: string, data: any) {
  return Mustache.render(template, data);
}

function emptyFn() {}
async function emptyAsyncFn() {}

export function renderOptions<O extends Record<string, any>>({
  options,
  project,
  workspace,
  schema
}: {
  options: O;
  project: Project;
  workspace: Workspace;
  schema: {
    properties?: {
      [prop: string]: {
        type?: string;
        format?: string;
        default?: string;
      };
    };
  };
}) {
  const templateData = {
    projectDir: project.fullPath,
    projectName: project.name
  };

  const schemaProperties = schema?.properties ?? {};

  const transformedOptions: Record<string, any> = {};
  Object.entries(schemaProperties).forEach(([propName, propDef]) => {
    const optionValue = options[propName] ?? propDef.default;
    transformedOptions[propName] =
      typeof optionValue === 'string' && propDef.format === 'path'
        ? (Path.resolve(
            workspace.cwd,
            renderTemplate(optionValue, templateData)
          ) as any)
        : optionValue;
  });
  return transformedOptions as O;
}

export function getRunnerContext<O extends Record<string, any>>(
  opts: GetRunnerContextOptions<O>
) {
  const {
    options,
    project,
    outputDir,
    workspace,
    input,
    fs = originalFs,
    dependsOnFile = emptyFn,
    watchDependencies = emptyFn,
    longRunning = emptyFn,
    commitFilesToFS = emptyAsyncFn,
    batch,
    logger,
    cacheProvider,
    defaultCacheKeys = []
  } = opts;

  function getContextBase(projectOptions: O, outputDir?: string) {
    return {
      options: {
        ...projectOptions,
        outputDir
      } as O & { outputDir?: string },
      renderTemplate
    };
  }

  const context = {
    ...getContextBase(options, outputDir),
    project,
    workspace,
    input,
    dependsOnFile,
    watchDependencies,
    longRunning,
    commitFilesToFS,
    *batch() {
      if (batch) {
        for (const { project, options } of batch) {
          yield {
            ...getContextBase(options),
            project
          };
        }
      } else {
        throw new Error('No batch');
      }
    },
    createOutputContainer(
      target: string | File,
      cacheKeys: string[],
      dependencies?: (string | File)[]
    ) {
      return new OutputContainer(
        cacheProvider,
        target,
        [...defaultCacheKeys, ...cacheKeys],
        dependencies
      );
    },
    fs,
    logger,
    file
  };

  return context;
}
