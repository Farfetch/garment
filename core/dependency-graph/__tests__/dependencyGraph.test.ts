import { Graph } from './../src/DependencyGraph';

async function delay(ms: number) {
  return new Promise(resolve => {
    setTimeout(() => {
      resolve();
    }, ms);
  });
}

function createGraph(connections: { [name: string]: string[] }) {
  const graph = new Graph<string>();

  for (const [key, value] of Object.entries(connections)) {
    graph.addNode(key);
    for (const dep of value) {
      if (!graph.hasNode(dep)) {
        graph.addNode(dep);
      }
      graph.addDependency(key, dep);
    }
  }

  return graph;
}

test('traverses graph parallelly', async () => {
  const graph = createGraph({
    n1: ['n2', 'n3', 'n4'],
    n2: ['n5'],
    n3: ['n5'],
    n4: ['n6']
  });
  const nodesOrder: string[] = [];

  await graph.traverseParallel(async node => {
    switch (node) {
      case 'n4':
        await delay(50);
        break;
      case 'n5':
        await delay(10);
        break;
      case 'n3':
        await delay(30);
        break;
      default:
        break;
    }
    nodesOrder.push(node);
  });

  expect(nodesOrder).toMatchSnapshot();
});

test('returns independent subgraphs', () => {
  const graph = createGraph({
    n1: ['n2', 'n3'],
    n2: ['n5'],
    n3: ['n5'],
    n4: ['n6'],
    n7: ['n4']
  });

  const subgraphs = graph.getIndependentSubgraphs();
  expect(subgraphs.map(_ => _.getOverallOrder())).toMatchSnapshot();
});

test('filters', () => {
  const graph = createGraph({
    n1: ['n2'],
    n2: ['n3']
  });

  const filtered = graph.filter(_ => _ !== 'n2');
  expect(filtered).toMatchSnapshot();
});
