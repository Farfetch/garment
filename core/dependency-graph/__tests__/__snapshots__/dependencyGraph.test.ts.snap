// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filters 1`] = `
Graph {
  "inEdges": Map {
    "n1" => Set {},
    "n3" => Set {},
  },
  "nodes": Set {
    "n1",
    "n3",
  },
  "outEdges": Map {
    "n1" => Set {},
    "n3" => Set {},
  },
}
`;

exports[`returns independent subgraphs 1`] = `
Array [
  Array [
    "n5",
    "n3",
    "n2",
    "n1",
  ],
  Array [
    "n6",
    "n4",
    "n7",
  ],
]
`;

exports[`traverses graph parallelly 1`] = `
Array [
  "n6",
  "n5",
  "n2",
  "n3",
  "n4",
  "n1",
]
`;
